lib/secretsscanner/reporter/report.go

```diff
@@ -0,0 +1,211 @@
+
+	"github.com/gravitational/trace"
+
+	accessgraphsecretsv1pb "github.com/gravitational/teleport/api/gen/proto/go/teleport/accessgraph/v1"
+	devicepb "github.com/gravitational/teleport/api/gen/proto/go/teleport/devicetrust/v1"
+	dtassert "github.com/gravitational/teleport/lib/devicetrust/assert"
+	secretsscannerclient "github.com/gravitational/teleport/lib/secretsscanner/client"
+)
+
+const (
+	defaultBatchSize = 100
+)
+
+// AssertCeremonyBuilderFunc is a function that builds the device authentication ceremony.
+type AssertCeremonyBuilderFunc func() (*dtassert.Ceremony, error)
+
+// Config specifies the configuration for the reporter.
+type Config struct {
+	// ProxyAddress is the address of the proxy server to send the report to.
+	ProxyAddress string
+	// InsecureSkipVerify specifies whether to trust the certificates from the Proxy Server during registration without verification.
+	InsecureSkipVerify bool
+	// Log is the logger.
+	Log *slog.Logger
+	// BatchSize is the number of secrets to send in a single batch. Defaults to [defaultBatchSize] if not set.
+	BatchSize int
+	// AssertCeremonyBuilder is the device authentication ceremony builder.
+	// If not set, the default device authentication ceremony will be used.
+	AssertCeremonyBuilder AssertCeremonyBuilderFunc
+}
+
+// Reporter reports secrets to the Teleport Proxy.
+type Reporter struct {
+	proxyAddr             string
+	insecureSkipVerify    bool
+	log                   *slog.Logger
+	batchSize             int
+	assertCeremonyBuilder AssertCeremonyBuilderFunc
+}
+
+// New creates a new reporter instance.
+func New(cfg Config) (*Reporter, error) {
+	if cfg.ProxyAddress == "" {
+		return nil, trace.BadParameter("missing proxy address")
+	}
+	if cfg.Log == nil {
+		cfg.Log = slog.Default()
+	}
+	if cfg.BatchSize == 0 {
+		cfg.BatchSize = defaultBatchSize
+	}
+	if cfg.AssertCeremonyBuilder == nil {
+		cfg.AssertCeremonyBuilder = func() (*dtassert.Ceremony, error) {
+			return dtassert.NewCeremony()
+		}
+	}
+	return &Reporter{
+		proxyAddr:             cfg.ProxyAddress,
+		insecureSkipVerify:    cfg.InsecureSkipVerify,
+		log:                   cfg.Log,
+		batchSize:             cfg.BatchSize,
+		assertCeremonyBuilder: cfg.AssertCeremonyBuilder,
+	}, nil
+}
+
+// ReportPrivateKeys reports the private keys to the Teleport server.
+// This function performs the following steps:
+// 1. Create a new gRPC client to the Teleport Proxy.
+// 2. Run the device assertion ceremony.
+// 3. Report the private keys to the Teleport cluster.
+// 4. Wait for the server to acknowledge the report.
+func (r *Reporter) ReportPrivateKeys(ctx context.Context, pks []*accessgraphsecretsv1pb.PrivateKey) error {
+	client, err := secretsscannerclient.NewSecretsScannerServiceClient(
+		ctx,
+		secretsscannerclient.ClientConfig{
+			ProxyServer: r.proxyAddr,
+			Insecure:    r.insecureSkipVerify,
+			Log:         slog.Default(),
+		})
+	if err != nil {
+		return trace.Wrap(err, "failed to create client")
+	}
```

Comment:
Should this receive a secretsscannerclient.Client instead of create its own? By doing that we wouldn't have to duplicate its config settings here, making the design more resilient and flexible.